gio/src/subclass/vfs.rs

```diff
@@ -0,0 +1,386 @@
+        klass.parse_name = Some(parse_name::<T>);
+    }
+}
+
+unsafe extern "C" fn is_active<T: VfsImpl>(vfs: *mut ffi::GVfs) -> glib::ffi::gboolean {
+    let instance = &*(vfs as *mut T::Instance);
+    let imp = instance.imp();
+
+    let res = imp.is_active();
+
+    res.into_glib()
+}
+
+unsafe extern "C" fn get_file_for_path<T: VfsImpl>(
+    vfs: *mut ffi::GVfs,
+    path: *const c_char,
+) -> *mut ffi::GFile {
+    let instance = &*(vfs as *mut T::Instance);
+    let imp = instance.imp();
+
+    let file = imp.get_file_for_path(&from_glib_borrow(path));
+
+    file.to_glib_full()
+}
+
+unsafe extern "C" fn get_file_for_uri<T: VfsImpl>(
+    vfs: *mut ffi::GVfs,
+    uri: *const c_char,
+) -> *mut ffi::GFile {
+    let instance = &*(vfs as *mut T::Instance);
+    let imp = instance.imp();
+
+    let file = imp.get_file_for_uri(&from_glib_borrow(uri));
+
+    file.to_glib_full()
+}
+
+unsafe extern "C" fn get_supported_uri_schemes<T: VfsImpl>(
+    vfs: *mut ffi::GVfs,
+) -> *const *const c_char {
+    let instance = &*(vfs as *mut T::Instance);
+    let imp = instance.imp();
+
+    let supported_uri_schemes = imp.get_supported_uri_schemes();
+
+    supported_uri_schemes.to_glib_full()
+}
+
+unsafe extern "C" fn parse_name<T: VfsImpl>(
+    vfs: *mut ffi::GVfs,
+    parse_name: *const c_char,
+) -> *mut ffi::GFile {
+    let instance = &*(vfs as *mut T::Instance);
+    let imp = instance.imp();
+
+    let file = imp.parse_name(&from_glib_borrow(parse_name));
+
+    file.to_glib_full()
+}
+
+#[cfg(test)]
+mod tests {
+    use std::cell::RefCell;
+    use std::path::PathBuf;
+
+    use glib::object::ObjectSubclassIs;
+
+    use super::*;
+    use crate::prelude::*;
+
+    mod imp {
+        use super::*;
+
+        #[derive(glib::Properties, Default)]
+        #[properties(wrapper_type = super::MyFile)]
+        pub struct MyFile {
+            #[property(get, set)]
+            path: RefCell<Option<std::path::PathBuf>>,
+        }
+
+        #[glib::object_subclass]
+        impl ObjectSubclass for MyFile {
+            const NAME: &'static str = "MyFile";
+            const ABSTRACT: bool = false;
+            type Type = super::MyFile;
+            type Interfaces = (File,);
+        }
+
+        #[glib::derived_properties]
+        impl ObjectImpl for MyFile {}
+
+        impl MyFile {
+            fn uri_scheme(&self) -> Option<glib::GString> {
+                Some(GString::from(MyVfs::NAME))
+            }
+
+            fn uri(&self) -> glib::GString {
+                GString::from(format!(
+                    "{}:{}",
+                    MyVfs::NAME,
+                    self.obj()
+                        .path()
+                        .unwrap_or(std::path::PathBuf::from("no_name"))
+                        .to_str()
+                        .unwrap()
+                ))
+            }
+        }
+
+        unsafe impl IsImplementable<MyFile> for File {
```

Comment:
Can you split this into subclass/file.rs even if it only implements a small subset of the vfuncs, the rest can be added later.